File: LeetCode/878.shifting-letters/shifting-letters.java
class Solution {
    public String shiftingLetters(String S, int[] shifts) {
        char[] str = new char[S.length()];
        int finalen = shifts[shifts.length - 1];
        for (int i = shifts.length - 2;i >= 0;i--){
            int tmp = shifts[i];
            shifts[i] = (shifts[i] + finalen) % 26;
            finalen = (finalen + tmp) % 26;
        }
        for (int i = 0;i < S.length();i++){
            str[i] = (char)((shifts[i] + S.charAt(i) - 97) % 26 + 97);
        }
        String s = new String(str);
        return  s;
    }
}

File: LeetCode/0848.shifting-letters/shifting-letters.cpp
class Solution {
public:
    string shiftingLetters(string S, vector<int>& shifts) {
        int finalen = shifts[shifts.size() - 1];
        for (int i = shifts.size() - 2;i >= 0;i--){
        	int tmp = shifts[i];
        	shifts[i] = (shifts[i] + finalen) % 26;
        	finalen = (finalen + tmp) % 26;
        }
        for (int i = 0;i < S.size();i++){
        	S[i] = (shifts[i] + S[i] - 97) % 26 + 97;
        }
        return S;
    }
};